Fix ignored entity pivot

Koopa1018 opened the pull request:
This PR modifies `calculate_transform_from_entity_instance` to use plain `ldtk_pixel_coords_to_translation` rather than the `pivoted` version. It also updates the function's test suite to reflect the different results it now gives.

Theoretically this PR will fix #290, but I've never worked with this codebase, so I have no idea if what I've done here is enough to actually fix the bug, nor do I know what else may need updating. (I know the enemies in the Platformer example show up displaced now.)

Koopa1018 committed:
Use un-pivoted conversion for entity translation

This is the basic ingredient needed to fix #290. There's a lot of things which will potentially break if this is done in isolation, so there's a lot more work ahead.

Koopa1018 committed:
Make to_translation_pivoted more readable

Comment it up, and also disambiguate use of the word "pivot."

It always helps to do this when I'm first trying to understand a codebase.

Koopa1018 committed:
Update tests to match new translation behavior

Feels kind of silly testing behavior that's just "match the coordinates and flip the Y," but then again, the scale is worth testing.

Trouv commented on src/utils.rs:
```diff
@@ -85,11 +85,9 @@ pub fn calculate_transform_from_entity_instance(
     let size = IVec2::new(entity_instance.width, entity_instance.height);
 
-    let translation = ldtk_pixel_coords_to_translation_pivoted(
+    let translation = ldtk_pixel_coords_to_translation(
```
This is definitely a part of the change that needs to happen, to merge this though this PR should replace this with the anchoring functionality. I believe this will require updating the `TextureAtlasSprite` construction in `sprite_sheet_bundle_from_entity_info` to use the `pivot` value of `&EntityInstance`

Koopa1018 replied:
Ahhh, in the sprite bundle...that makes sense. Glad you reminded me, that would've been a stupid thing to miss 🤦

Trouv reviewed (changes requested):
Thanks for the contribution!

> nor do I know what else may need updating

No worries. The examples definitely should be updated to match their previous behavior.

Also, since this is a breaking change, a new section in the migration guide is in order (at `book/src/how-to-guides/migration-guides/migrate-from-0.8-to-0.9.md`)

Koopa1018 committed:
Use if-let in sprite_sheet_bundle_from_entity_info

Match usually implies multiple values need to be handled explicitly. When it's like this--everything or nothing--an if let expresses the intent better.

Koopa1018 committed:
Set sprite anchor from entity pivot

Now *this* is how the pivot should be used! Not placing the entity--placing the sprite *relative* to the entity :)

Koopa1018 committed:
Fix traitless example

The whole point of this one is that it doesn't use the code I fixed. Have to apply the fix manually, then. (Confirmed working now.)

Koopa1018 commented:
Arright, I've done what I can with the examples. Most appear fine, and `traitless` was easy enough to fix, but I have no idea where to even look in the platformer example....
My only leads are that when the example begins running, the skulls slowly rise into the air as they move, until they reach this height: 
![image](https://github.com/Trouv/bevy_ecs_ldtk/assets/29995463/e58898ba-e474-400d-9aef-2c43587b008d)
and that the boss (who does not move) has a lowered hitbox:
![image](https://github.com/Trouv/bevy_ecs_ldtk/assets/29995463/1b80f218-a349-45f8-807c-c78f89ca5ca3)

Migration guide is going smoothly. Not totally sure how to demonstrate the changed translation/pivot behavior in a code example, though....

Also: I notice the CI format check failed. I've never contributed to a project with CI, so I'm not sure what that means, even after checking the details. Are these formatting issues that I need to fix?

Koopa1018 committed:
Dedicated fn for finding entity center

To ease migration process, a function which produces the same value as translation() did before the anchor/pivot fix.

Koopa1018 committed:
First draft of pivot behavior migration guide

This is a breaking change, and thus must be noted in the 0.8-0.9 migration guide. Draft because I can't quite figure out what to put in the first example.

Trouv commented:
> Also: I notice the CI format check failed. I've never contributed to a project with CI, so I'm not sure what that means, even after checking the details. Are these formatting issues that I need to fix?

Yeah it'll need to be fixed before this is merged. For the format check in particular, it's just a matter of running `rustfmt`: https://github.com/rust-lang/rustfmt

Though it's something small enough that I'd do it myself if that was the only issue preventing merge.

BTW, some unsolicited advice: I recommend setting up your editor to automatically run rustfmt on save. It's so handy writing a ton of dirty code then hitting save and watching it prettify before your eyes

Koopa1018 committed:
rustfmt my changes

I'm looking at what it did so I can update my coding style. I'm seeing:
- Rust Analyzer's in-line type display misled me about how long line 88 and fn entity_center's definition were. They did not need to be shortened.
- Indents -> spaces. (Imma let rustfmt handle that one :sweat_smile:)
- Ah, so *that's* how you split long if-lets up...
- Do indent end-of-line comments to the same point. (In GDScript that's a no-no, so I'm used to avoiding that....)

Koopa1018 commented:
Can I get some help with the platformer example? It seems like to get the colliders to spawn right, it would require wider-reaching redesigns than I signed on for (that or I'm just totally missing a property that would fix everything...).

Trouv commented on book/src/how-to-guides/migration-guides/migrate-from-0.8-to-0.9.md:
````diff
@@ -429,3 +429,43 @@ let level_set = LevelSet::from_iids(
 ```
+
+## Entity translations now respect pivot
+Entities now spawn with a translation matching their pivot, rather than their center point.
````
I think implying that the new translation is their "actual position" whereas the old behavior is their "visual center" would be a little more clear
```suggestion
Entities now spawn with a translation matching their actual location, rather than their "visual center".
```

Trouv commented on book/src/how-to-guides/migration-guides/migrate-from-0.8-to-0.9.md:
````diff
@@ -429,3 +429,43 @@ let level_set = LevelSet::from_iids(
+// Demonstrate loading an entity in 0.8, assert the position.
+// Then, do the same in 0.9, demonstrating that it's not the same.
+```
+The entities' sprites calculate their `Anchor` from the pivot, so they should appear the same,
````
This is good but it should qualify this w/ the fact that this is only for sprite sheet bundles generated w/ the `#[sprite_sheet_bundle(...)]` attribute of `LdtkEntity`. Something like...
```suggestion
For `LdtkEntity` bundles with a `#[sprite_sheet_bundle(...)]`,  the macro calculates the sprite's `Anchor` from the pivot, so they should appear the same,
```

Trouv commented on book/src/how-to-guides/migration-guides/migrate-from-0.8-to-0.9.md:
````diff
@@ -429,3 +429,43 @@ let level_set = LevelSet::from_iids(
+		//do something with the center
+	}
+}
+```
````
Yeah it's hard to say what the best transition strategy is. I think for things like rapier colliders. most users should probably make the collider a "child" of their entity rather than the main entity. Or, if they don't want to deal with that, they could make the entity's pivot center.

This goes for the platformer example as well.. but it's a big change so I completely understand if you're not interested in doing all that work. Don't worry, wherever you feel like leaving off on this PR I can pick it up (thanks for making it maintainer-editable)

Trouv commented on examples/traitless.rs:
```diff
@@ -51,6 +51,10 @@ fn process_my_entity(
                     index: (tile.y / tile.h) as usize * 16 + (tile.x / tile.w) as usize,
+                    anchor: bevy::sprite::Anchor::Custom(Vec2::new(
+                        entity_instance.pivot.x - 0.5,
+                        0.5 - entity_instance.pivot.x,
```
Ah this is interesting. Maybe having a utility like `fn ldtk_pivot_to_anchor(pivot: Vec2) -> Anchor` would be helpful.

Koopa1018 replied:
Ooh, good thought.

Trouv commented on src/utils.rs:
```diff
@@ -227,13 +222,26 @@ pub fn ldtk_pixel_coords_to_translation_pivoted(
+    translation + offset
+}
+
+pub fn entity_center(translation: Vec3, entity_size: IVec2, pivot: Vec2) -> Vec3 {
```
I think `ldtk_entity_visual_center` could be a more consistent name for this.

Trouv replied:
Could you leave a brief doc for this? Something like...
```suggestion
/// Calculate the "visual center" of an  LDtk object using its current bevy translation, size, and pivot value.
pub fn entity_center(translation: Vec3, entity_size: IVec2, pivot: Vec2) -> Vec3 {
```

Trouv commented on src/utils.rs:
```diff
@@ -227,13 +222,26 @@ pub fn ldtk_pixel_coords_to_translation_pivoted(
+    let offset = entity_size.as_vec2() * adjusted_pivot;
+
+    translation + offset.extend(0.0)
 }
```
Also, some simple test cases for this would be amazing

Trouv reviewed (commented):
Thanks again for your work on this. The scope has definitely increased a lot, sorry about that.

One of the comments in this review does address a particular strategy for updating the platformer example. However, it's a lot of work, so I don't expect you to do it necessarily. I can happily pick up wherever you leave off.

Koopa1018 commented on examples/traitless.rs:
```diff
@@ -51,6 +51,10 @@ fn process_my_entity(
                     index: (tile.y / tile.h) as usize * 16 + (tile.x / tile.w) as usize,
+					anchor: bevy::sprite::Anchor::Custom(Vec2::new(
+						entity_instance.pivot.x - 0.5,
+						0.5 - entity_instance.pivot.x,
```
Oh frick yes it absolutely does 🤦

Koopa1018 committed:
Actually use pivot.y for the Y axis

Hate it when I miss this.

Koopa1018 committed:
Use clearer wording in pivot migration guide

Co-authored-by: Trevor Lovell <trevorlovelldesign@gmail.com>

Koopa1018 committed:
Clarify migration guide further

The anchor calculation depends on using the macro, as Trouv pointed out. And looking at my work later, I figured out how to make the center function clearer.

Co-Authored-By: Trevor Lovell <trevorlovelldesign@gmail.com>

Koopa1018 committed:
Make pivot->anchor conversion a fn

Because I can't imagine doing the conversions by hand every time would be at all easy on my sanity.

Replace both existing instances of this code with calls to the new function, too, because code sharing is tight.

Co-Authored-By: Trevor Lovell <trevorlovelldesign@gmail.com>

Koopa1018 committed:
Mention pivot->anchor fn in migration guide

Doesn't do anyone any good unless they know it's there (or can guess, which, you never know...).

Koopa1018 committed:
Rename entity_center for style consistency

Co-Authored-By: Trevor Lovell <trevorlovelldesign@gmail.com>

Koopa1018 committed:
Document center-finding function

Co-authored-by: Trevor Lovell <trevorlovelldesign@gmail.com>

Koopa1018 committed:
Merge branch 'fix/entity-pivot' of https://github.com/Koopa1018/bevy_ecs_ldtk into fix/entity-pivot

Koopa1018 commented on book/src/how-to-guides/migration-guides/migrate-from-0.8-to-0.9.md:
````diff
@@ -429,3 +429,43 @@ let level_set = LevelSet::from_iids(
+// Must define pivot first.
+// Demonstrate loading an entity in 0.8, assert the position.
+// Then, do the same in 0.9, demonstrating that it's not the same.
+```
````
Screenshots! That's perfect. Clearer than any text example I can think of....

Where should I store the actual image? Should I put it in `/repo/` with the readme image, or should it go next to the migration guides? Possibly in its own folder?

Trouv replied:
Yeah put it in a new directory next to the migration guides. Check out how it's done for the "tile based game" tutorial: https://github.com/Trouv/bevy_ecs_ldtk/blob/main/book/src/tutorials/tile-based-game/create-your-ldtk-project.md

there's an `images` directory there and the markdown is pointing to it w/ a title and relative path like:
```
![world-layout](images/world-layout.png)
```

Koopa1018 replied:
Sweet, thanks. Should've known there was a better example to check!

Koopa1018 committed:
Add diagram of pivot behavior between versions

For now, I'm putting the file in repo with the other image, knowing that it can be moved later.

The image itself needs to be updated for clarity--I tested on my brother, and he was so distracted by the bright shiny object that he totally missed the tiny pivot dot.

Koopa1018 committed:
Apply pivot to non-sheet sprite bundle too

Completely overlooked this function! Yes, these should probably also respect the pivot point.

Understanding the macro it's used in took a bit, but thankfully the change wasn't too hard to apply once I did.

Koopa1018 committed:
Rustfmt today's changes

More indent->space fixes. I should probably change that in VSCode.

Koopa1018 committed:
Simple test suite for ldtk_pivot_to_anchor

Asserts that
1. each axis is remapped from (0 - 1) to (-0.5 - 0.5) as is required for Bevy
2. Y is properly inverted.

Koopa1018 committed:
Rustfmt new test suite

uh...ok? Clearing out unneeded whitespace from a blank line....

Koopa1018 committed:
Move pivot diagram next to migration guide

Glad to find there actually was precedent for this, found in the book's tutorials folder.

Koopa1018 committed:
Clarify pivot diagram

Less contrasting entity graphic, bigger and more contrasting position reticule.

Koopa1018 committed:
Move pivot migration guide to new document

Missed 0.9, so it's gonna have to be in 0.10.

Koopa1018 committed:
Update pivot diagram for version differences

Koopa1018 commented:
Alright, I've done what I can (including updating the migration guide to the new version). The rest is in your hands.... See you in 0.10 🖖

Koopa1018 committed:
Merge remote-tracking branch 'upstream/main' into fix/entity-pivot

Koopa1018 committed:
Merge branch 'main' into fix/entity-pivot

Koopa1018 committed:
Reorganize platformer example

THIS. EXAMPLE. IS. A MESS! How can anyone learn from a haphazard jumble like this? It's like grabbing an armful of notes from a physicist's desk and calling it a textbook....

Anyway. I need to be able to sort through this, so I just decided to go ape. No logic was changed--it's just organized intuitively now.

Koopa1018 commented:
Arright, so I reorganized the platformer example so I could make heads or tails of it. Now there's a concrete problem in front of me: I'm not sure how to spawn colliders as children within the `LdtkEntity` framework. If I had an `Entity` ref, no problem, but...

Think I'm gonna have to leave this as draft until #177 finally hits. Until then, might as well submit the reorganized platformer example as a separate PR so it doesn't become hell to merge down the road....

Trouv commented:
Thanks for all your hard work on this. I have a lot more time for this project lately and I'm still interested in getting this PR in. Sorry for the wait.

It could use a rebase, your platformer refactor _has_ gotten in in the meantime, and some of the migration guide stuff will need to refer to newer versions of the crate (0.13 and 0.14 instead of 0.9 and 0.10).

> Now there's a concrete problem in front of me: I'm not sure how to spawn colliders as children within the LdtkEntity framework. If I had an Entity ref, no problem, but...

I'm not 100% sure what this has to do with this PR, is it a blocker? normally for things like this you do have to do post-processing instead of just `LdtkEntity` + macros

Koopa1018 committed:
Merge branch 'main' into fix/entity-pivot

with some rather intense custom merging 😖 Some spots need further editing to reimplement their ldtk_pivot_to_anchor calls...some of those edits rather structural in nature. Would rather finish the merge first before beginning on those....

Koopa1018 committed:
Fix the easy pivot-to-anchor insertion

One down...two breaking API changes to go.

Koopa1018 committed:
Restore utils.rs functions lost in merge

In such a big merge, it's no wonder things were lost.

Koopa1018 committed:
Post-process system to assign sprite anchors

Feels like a stopgap fix, an ugly hack that should work better.... Would love to write a test suite for this, but I don't know where I'd put that.

Koopa1018 committed:
Delete physics.rs

Holy heck this never got removed. I split its functionality out into `ground_detection` and `colliders`, but...I guess the file itself never got removed?

Koopa1018 committed:
Remove anchor postpro, manual anchor in Traitless

I don't think this post-process system is It. Even if it is, traitless.rs is about stepping *outside* of this system. Actually, traitless.rs is what clued me in that this method might be wrong....

Koopa1018 committed:
Fix function invocation on #[sprite] fields

Forgot to remove the entity_instance reference from this invocation when I removed it from the signature.

Koopa1018 committed:
Load anchor during level loading

There we go! Adds anchor automatically, without making the user opt in, but lets the user override the Anchor if wanted.

Update LdtkEntity::bundle_entity documentation to explain how Anchor is added, and that it can be overridden. Trouv says we may want to move Transform to being added pre-bundle as well so it can be overridden, but definitely not in this PR.

Koopa1018 committed:
Restore all pivot-related changes to utils.rs

Well that's infuriating. Looks like merging from main undid most of my changes to this file 😖 Let's get that back in order....

The examples that existed when I started (platformer notwithstanding) are all in working order now. I don't remember tile_based_game or collectathon...may need to update those ones.

Koopa1018 committed:
Fix translations in tile_based_game

The pivot code invalidates this example's assumption that the entity position is at the center of the grid.

I could fix this by changing the pivots in the LDtk project...but it seems better to demonstrate how to adjust it when the pivot is wrong. I want to fix it properly by making a pivot-aware grid-position function, but that's a fair bit more complex than I can handle tonight...might even be impossible to handle in a generally-correct way.

Update code references in add-gameplay-to-your-project.md as well. Need to update the textual content of this document to address the changed behavior.
